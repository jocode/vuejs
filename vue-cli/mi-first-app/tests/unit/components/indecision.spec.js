import { shallowMount } from '@vue/test-utils'
import Indecision from '@/components/Indecision.vue'

describe('Indecision Component', () => {

    let wrapper
    let consoleLogSpy

    // Mock del 'fetch' Api
    global.fetch = jest.fn( () => Promise.resolve({
        json: () => Promise.resolve({
            answer: "yes",
            forced: false,
            image: "https://yesno.wtf/assets/yes/2.gif"
        })
    }))

    beforeEach(() => {
        wrapper = shallowMount(Indecision)
        consoleLogSpy = jest.spyOn(console, 'log')

        jest.clearAllMocks()
    })

    test('Debe hacer match con el snapshot', () => {
        expect(wrapper.html()).toMatchSnapshot()
    })

    test('escribir en el input no debe de disparar nada (console.log)', async() => {

        const getAnswerSpy = jest.spyOn(wrapper.vm, 'getAnswer') // vm es la instancia del Vue

        const input = wrapper.find('input')
        await input.setValue('Hola mundo')

        // expect(consoleLogSpy).toHaveBeenCalled()
        expect(consoleLogSpy).toHaveBeenCalledTimes(1)
        expect(getAnswerSpy).toHaveBeenCalledTimes(0)
        // expect(getAnswerSpy).not.toHaveBeenCalled()

    })

    test('escribir el simbolo de "?" debe disparar el fetch', async() => {
        const getAnswerSpy = jest.spyOn(wrapper.vm, 'getAnswer')

        const input = wrapper.find('input')
        await input.setValue('Te gusta Vue?')

        expect(consoleLogSpy).toHaveBeenCalledTimes(1)
        expect(getAnswerSpy).toHaveBeenCalledTimes(1)
    })

    test('pruebas en getAnswer', async() => {
        await wrapper.vm.getAnswer()
        // console.log(wrapper.vm.img)
        // console.log(wrapper.vm.answer)

        const img = wrapper.find('img')
        
        expect( img.exists() ).toBeTruthy()
        expect( wrapper.vm.img ).toBe('https://yesno.wtf/assets/yes/2.gif')
        expect(wrapper.vm.answer).toBe('Â¡Si!')
    })

    test('pruebas en getAnswer - Fallo en el API', async() => {
        // TODO: Failed API
        fetch.mockImplementationOnce( () => Promise.reject('API is down'))

        await wrapper.vm.getAnswer()

        const img = wrapper.find('img')
        expect( img.exists() ).toBeTruthy()
        expect( wrapper.vm.answer ).toBe('No se pudo cargar del API')
    })

})