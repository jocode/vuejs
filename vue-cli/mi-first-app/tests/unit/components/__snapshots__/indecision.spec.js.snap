// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Indecision Component Debe hacer match con el snapshot 1`] = `
<img src="https://picsum.photos/seed/picsum/800/600" alt="Imagen">
<div class="bg-dark"></div>
<div class="indecision-container">
  <p>Recuerda terminar con un signo de interrogaci√≥n (?)</p><input type="text" placeholder="Hazme una pregunta">
  <!--v-if-->
</div>
`;
